runner/host/processes.go

```diff
@@ -22,11 +32,50 @@ func (p Process) Run() op.Op {
 		return op.New(p.ID(), processes, op.Fail, err, nil)
 	}
 
-	processInfo := make([]string, 0)
-	for eachProcess := range processes {
-		process := processes[eachProcess]
-		processInfo = append(processInfo, process.Executable())
+	// Maps parent PIDs to child processes
+	var processInfo = make(map[int][]proc)
+
+	for _, process := range processes {
+		newProc := proc{
+			Name: process.Executable(),
+			PID:  process.Pid(),
+			PPID: process.PPid(),
+		}
+		// Append to slice of Process under this proc's PPID
+		processInfo[newProc.PPID] = append(processInfo[newProc.PPID], newProc)
+	}
+
+	// Reprocess our original processInfo map with the process list, to populate parent names in our new namedProcessMap
+	// (only a single pass through the process list each time, to avoid using quadratic time)
+	var namedProcessMap = make(map[string][]proc)
+
+	for _, process := range processes {
+		// If our PID is listed as a PPID in processInfo, we're a parent!
+		children, ok := processInfo[process.Pid()]
+		if ok {
+			newProc := proc{
+				Name: process.Executable(),
+				PID:  process.Pid(),
+				PPID: process.PPid(),
+			}
+
+			// Process names are not unique by default, but we need unique keys
+			uniqueName := uniqueProcName(newProc)
+			namedProcessMap[uniqueName] = children
+		}
+	}
+
+	return op.New(p.ID(), namedProcessMap, op.Success, nil, nil)
+}
+
+// Creates a unique name from a proc's name and PID
+func uniqueProcName(process proc) string {
```

Comment:
Should this take a `*proc` instead?

Reply:
If the state of `process` isn't being changed, a value type is probably better than a pointer. Pointers run the risk of side-effects, so if none are intended, then a value is usually best. (Exceptions can be made for very large objects, where passing a pointer might be enough of a memory improvement to warrant it. Although pointers also are more likely to wind up on the heap, which causes work for the garbage collector... Everything's a tradeoff. 😄 )

All that said, this MIGHT be a case where we should make this a method with a `proc` receiver. This function is unexported, so only used in the `runner/host` package, and it operates exclusively on a `proc`. So, perhaps it could become something like:

```go
func (p proc) uniqueName() string {
    pl := len(p.Name)
    // ... 
}
```

Then, lines 63-64 could become something like:

```go
namedProcessMap[newProc.uniqueName()] = children
```

Reply:
Thanks, that totally makes sense -- I was definitely thinking of optimizing away that tiny bit of memory usage for the value type, which is silly. And you're right, `uniqueProcName()` is very much "about" the proc, so I like the idea of giving it a proc receiver. I was trying to err on the side of avoiding things that looked like old OO habits 😄 .

Appreciate it!